e2e/specs/st_file_uploader.spec.js

```diff
@@ -393,4 +393,51 @@ describe("st.file_uploader", () => {
       );
     });
   });
+
+  // regression test for https://github.com/streamlit/streamlit/issues/4256 bug
+  it("not calls callback if not changed", () => {
+    const fileName1 = "file1.txt";
+    const uploaderIndex = 4;
+
+    cy.fixture(fileName1).then(file1 => {
+      const files = [
+        { fileContent: file1, fileName: fileName1, mimeType: "text/plain" }
+      ];
+
+      // Script contains counter variable stored in session_state with
+      // default value 0. We increment counter inside file_uploader callback
+      // Since callback did not called at this moment, counter value should
+      // be equal 0
+      cy.getIndexed("[data-testid='stText']", uploaderIndex).should(
+        "contain.text",
+        "0"
+      );
+
+      // Uploading file, should invoke on_change call and counter increment
+      cy.getIndexed(
+        "[data-testid='stFileUploadDropzone']",
+        uploaderIndex
+      ).attachFile(files[0], {
+        force: true,
+        subjectType: "drag-n-drop",
+        events: ["dragenter", "drop"]
+      });
+
+      // Make sure callback called
+      cy.getIndexed("[data-testid='stText']", uploaderIndex).should(
+        "contain.text",
+        "1"
+      );
+
+      // On rerun, make sure callback is not called, since file not changed
+      cy.get("body").type("r");
+      cy.wait(1000);
```

Comment:
Not necessarily something to stop the PR for, but I wonder if there's a more deterministic way to wait for the rerun to happen than `cy.wait(1000)`. Could we instead test for the presence of the "Running..." indicator?

Reply:
+1

The newly-added `loadApp` custom cypress command (in `frontend/cypress/support/commands.js`) has an example of how to do this. It sounds like it may be useful to generalize that into a a cypress command so that we have a standard way of waiting until an app has finished running